File: lib/fr/minuskube/inv/SmartInventory.ts
declare var Java: any;
import {SmartInventory$Builder} from '../../../fr/minuskube/inv/SmartInventory$Builder.js'
import {Player} from '../../../org/bukkit/entity/Player.js'
import {Inventory} from '../../../org/bukkit/inventory/Inventory.js'
import {InventoryType} from '../../../org/bukkit/event/inventory/InventoryType.js'
import {InventoryProvider} from '../../../fr/minuskube/inv/content/InventoryProvider.js'
import {InventoryManager} from '../../../fr/minuskube/inv/InventoryManager.js'

export interface SmartInventory {
	getParent(): any;
	getId(): string;
	close(arg0: Player): void;
	open(arg0: Player, arg1: number): Inventory;
	open(arg0: Player): Inventory;
	getType(): InventoryType;
	getProvider(): InventoryProvider;
	getRows(): number;
	isCloseable(): boolean;
	setCloseable(arg0: boolean): void;
	getTitle(): string;
	getManager(): InventoryManager;
	getColumns(): number;
}

export class SmartInventory {
	public static get $javaClass(): any {
		return Java.type('fr.minuskube.inv.SmartInventory');
	}
	public static builder(): SmartInventory$Builder;
	public static builder(...args: any[]): any {
		return SmartInventory.$javaClass.builder(...args);
	}
}

File: ScriptableMC-Engine-JS/src/main/kotlin/com/smc/smartinvs/SmartInventory.kt
package com.smc.smartinvs

import com.pixlfox.scriptablemc.core.ScriptablePluginEngine
import fr.minuskube.inv.ClickableItem
import fr.minuskube.inv.SmartInventory
import fr.minuskube.inv.content.InventoryContents
import fr.minuskube.inv.content.InventoryProvider
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemStack
import org.graalvm.polyglot.Value
import java.util.function.Consumer

@Suppress("unused", "MemberVisibilityCanBePrivate")
class SmartInventory {
    companion object {
        @JvmStatic
        fun builder(): SmartInventory.Builder = SmartInventory.builder().manager(ScriptablePluginEngine.instance!!.inventoryManager)

        @JvmStatic
        fun provider(scriptableObject: Value): SmartInventoryProvider {
            if (scriptableObject.canInstantiate()) {
                return provider(scriptableObject.newInstance())
            }

            return SmartInventoryProvider(scriptableObject)
        }

        @JvmStatic
        fun clickableItem(item: ItemStack): ClickableItem = ClickableItem.empty(item)

        @JvmStatic
        fun clickableItem(item: ItemStack, consumer: Consumer<InventoryClickEvent>): ClickableItem = ClickableItem.of(item, consumer)
    }
}

class SmartInventoryProvider(private val scriptableObject: Value) : InventoryProvider {
    override fun init(player: Player, contents: InventoryContents) {
        if(scriptableObject.hasMember("init") && scriptableObject.canInvokeMember("init")) {
            scriptableObject.invokeMember("init", player, contents)
        }
    }

    override fun update(player: Player, contents: InventoryContents) {
        if(scriptableObject.hasMember("update") && scriptableObject.canInvokeMember("update")) {
            scriptableObject.invokeMember("update", player, contents)
        }
    }
}